pxi.py

```diff
@@ -7,33 +7,35 @@
 responses from hardware to CsPy. 
 """
 
-'''
+"""
 general TODOs and ideas:
+- reset_connection never gets used here, as far as I can tell. maybe TCP references it?
+- there should be a way to re-start the measurement loop after an error is hit.
+the best way might be to close the thread running it and call launch_experiment_thread
+again 
 - could decorate methods like is_done with a timeout method
-- could implement error handling and checking whether connection has been 
-    stopped or restarted with a decorator
-- error object to bundle exception/error message returned with a boolean? 
-    would be useful in loops that should exit on exception raised
-'''
+"""
 
-#### modules
+## modules
 import socket
 import logging
 import threading
 import xml.etree.ElementTree as ET
 from typing import Tuple
 from queue import Queue, Empty
+from time import perf_counter_ns
 
-#### misc local classes
+## misc local classes
 from keylistener import KeyListener
+from pxierrors import XMLError, HardwareError, TimeoutError
 
-#### local device classes
+## local device classes
 from hsdio import HSDIO, HSDIOError
 from hamamatsu import Hamamatsu, IMAQError
 from tcp import TCP
 
 
-# TODO : Should this inherit from XMLLoader?
+# TODO : Should this inherit from XMLLoader? <-- i'm on board with this. - Preston
```

Comment:
What utility does that provide?

Reply:
Very little (logging would be set up in the super call, and the intended use of load_xml is essentially that of parse_xml). I think mostly it would just signal to readers that the spinal cord pxi is the same as an instrument. But upon second glance, XMLLoader would have to be modified slightly, and to your point, it's not worth it for now.